File: piece_recognition.py
"""
author: EMNEM
date-created: 10/2/22
last-modified: 10/2/22

piece_recognition.py
This file contains helper methods for detecting checkers pieces
"""

import cv2
import numpy as np
import time

# The max distance in pixels squared
CLUSTER_MAX_RANGE = 10**2

# Contains info about group of pixels and how to add new ones
# point is tuple of (x, y, type), where 0 = blue, 1 = red, 2 = yellow
class Cluster:
    def __init__(self):
        self.red = 0
        self.blue = 0
        self.yellow = 0
        self.is_blue = None
        self.is_king = None
        self.points = []
        self.x_sum = 0
        self.y_sum = 0
        self.x = None
        self.y = None
        self.is_valid = False
        return
    
    # Adds point to self
    def add_point(self, point):
        # Update sumes
        self.points.append(point)
        self.x_sum += point[0]
        self.y_sum += point[1]
        # Check color
        if point[2] == 0:
            self.blue += 1
        elif point[2] == 1:
            self.red += 1
        else:
            self.yellow += 1
        return
    
    # checks of point is within range
    # returns True or False depending on whether close point found
    def check_range(self, p1):
        # Check for every point, break once a close one is found
        found = False
        for p2 in self.points:
            # Check colors match
            if p2[2] == 2 or p1[2] == 2 or p2[2] == p1[2]:
                # Check colors dont match
                if (p2[0] - p1[0])**2 + (p2[1] - p1[1])**2 < CLUSTER_MAX_RANGE:
                    # Point is close
                    self.add_point(p1)
                    found = True
                    break
        return found
    
    # finalizes cluster and dtermines center and type
    # returns true if cluster is likely a piece
    def finalize(self):
        total = self.red + self.blue + self.yellow
        if total < 5:
            # Not enough points for piece to exist
            return False
        # Check color
        if self.red > self.blue:
            self.is_blue = False
            if self.red < 10:
                # Not enough colored points
                return False
        else:
            if self.blue < 10:
                # Not enough colored points
                return False
            self.is_blue = True
        if self.yellow > 10:
            self.is_king = True
        else:
            self.is_king = False
        # Calc center
        self.x = int(self.x_sum / total)
        self.y = int(self.y_sum / total)
        self.is_valid = True
        return True

# Returns list of points of color in image
def get_points_in_img(img, scale=1):
    # Split and filter
    blue_img = img[:,:,0].astype(np.int16)
    green_img = img[:,:,1].astype(np.int16)
    red_img = img[:,:,2].astype(np.int16)
    blue_filter = blue_img - green_img - red_img
    print(blue_filter)
    red_filter = red_img - blue_img - green_img
    yellow_filter = 0.5*(red_img + green_img) - 2*blue_img
    # Find points
    kernel_size = 4//scale
    kernel_area = kernel_size ** 2
    filter_cutoff = 160 * kernel_area
    blue_points = []
    red_points = []
    for i in range(0, 1088, kernel_size):
        for j in range(0, 1920, kernel_size):
            # Check for blue
            if blue_filter[i:i+kernel_size,j:j+kernel_size].sum() > filter_cutoff:
                blue_points.append((i+kernel_size//2, j+kernel_size//2, 0))
            elif red_filter[i:i+kernel_size,j:j+kernel_size].sum() > filter_cutoff:
                red_points.append((i+kernel_size//2, j+kernel_size//2, 1))
            elif yellow_filter[i:i+kernel_size,j:j+kernel_size].sum() > filter_cutoff:
                blue_points.append((i+kernel_size//2, j+kernel_size//2, 2))
                red_points.append((i+kernel_size//2, j+kernel_size//2, 2))
    return blue_points, red_points

# turns list of points into list of clusters
def clusterize(p_list, is_blue):
    # create clusters as linked list
    clusters = []
    for point in p_list:
        # Go through all clusters and stop after finding one
        found = False
        for c in clusters:
            if c.check_range(point):
                c.add_point(point)
                found = True
                break
        if not found:
            # Create new cluster
            new_cluster = Cluster()
            new_cluster.add_point(point)
            clusters.append(new_cluster)
    # Finish cluster calculation
    final_clusters = []
    for c in clusters:
        c.finalize()
        if c.is_valid and c.is_blue == is_blue:
            # Probably a piece
            final_clusters.append(c)
    return final_clusters

# Test methods if run directly
if __name__ == "__main__":
    import sys
    start = time.perf_counter()
    # Check arguments
    if len(sys.argv) != 2:
        print("piece_recognition.py expects exactly one argument: the file name or 'pic'")
        exit()
    filename = sys.argv[1]
    img = None
    if filename == "pic":
        img = None
    else:
        img = cv2.imread(filename)
    img = img.astype(np.uint16)
    # img = cv2.resize(img, (0,0), fx=0.5, fy=0.5)
    # Check to make sure image is valid
    if img is None:
        print(f"File could not be read: {filename}")
        exit()
    # Start analyzing image
    blue_points, red_points = get_points_in_img(img, scale=1)
    points_end = time.perf_counter()
    print(f"# of blue: {len(blue_points)}, # of red: {len(red_points)}")
    blue_clusters = clusterize(blue_points, True)
    red_clusters = clusterize(red_points, False)
    print(f"There are {len(blue_clusters)} blue clusters and {len(red_clusters)} red clusters")
    end = time.perf_counter()
    """
    print("blue:")
    for c in blue_clusters:
        print(c.x, c.y)
    print("red:")
    for c in red_clusters:
        print(c.x, c.y)
    """
    print(f"Took {points_end - start:.3f} seconds for points")
    print(f"Took {end - points_end:.3f} seconds to clusterize")
    print(f"Took {end - start:.3f} seconds total")
    

File: PieceRecognition.cpp
/**
 * @file PieceRecognition.cpp
 * @author EMNEM
 * @brief 
 * @version 0.1
 * @date 2022-10-02
 * 
 * C++ implementation of piece_recognition.py
 */

#include <vector>
#include <iostream>
#include <opencv2/opencv.hpp>
#include "PieceRecognition.h"

//////////////////////////////// Cluster Definitions /////////////////////////////////////
Cluster::Cluster() {
    red = 0;
    blue = 0;
    yellow = 0;
    isBlue = false;
    isKing = false;
    isValid = false;
    xSum = 0;
    ySum = 0;
    x = 0;
    y = 0;
}

void Cluster::addPoint(Point& p) {
    // Add point to list and update sums
    points.push_back(p);
    xSum += p.x;
    ySum += p.y;
    // Check color and add to appropriate counter
    if(p.type == RED) {
        red += 1;
    }
    else if(p.type == BLUE) {
        blue += 1;
    }
    else {
        yellow += 1;
    }
}

bool Cluster::checkRange(Point& p1) {
    bool found = false;
    // Check for close points until one is found
    // BREAKS WHEN POINT IS FOUND
    for(Point& p2 : points) {
        // Check if points are same color or one is yellow
        if(p2.type == YELLOW || p1.type == YELLOW || p2.type == p1.type) {
            // Check distance
            int xDiff = p2.x - p1.x;
            int yDiff = p2.y - p1.y;
            if(xDiff*xDiff + yDiff*yDiff < CLUSTER_MAX_DISTANCE_SQUARE) {
                // Point is within range
                addPoint(p1);
                found = true;
                break;
            } 
        }
    }
    // Return whether close point was found
    return found;
}

bool Cluster::finalize() {
    // Check total number of points
    int totalPoints = red + blue + yellow;
    if(totalPoints < CLUSTER_MIN_POINTS) {
        // Not enough points to be piece
        return false;
    }
    // Check color
    if(red > blue) {
        isBlue = false;
        if(red < CLUSTER_MIN_POINTS) {
            // Not enough red points to be piece
            return false;
        }
    }
    else {
        isBlue = true;
        if(blue < CLUSTER_MIN_POINTS) {
            // Not enough blue points to be piece
            return false;
        }
    }
    // Check for king
    if(yellow >= CLUSTER_KING_MIN_POINTS) {
        isKing = true;
    }
    else {
        isKing = false;
    }
    // Calculate center
    x = xSum / totalPoints;
    y = ySum / totalPoints;
    isValid = true;
    return true;
}

//////////////////////////////// ImageState Definitions /////////////////////////////////////
int ImageState::countRedKingsOnBoard() {
    int counter = 0;
    for(CheckersPiece& piece : redPiecesOnBoard) {
        if(piece.isKing) {
            counter++;
        }
    }
    return counter;
}

int ImageState::countBlueKingsOnBoard() {
    int counter = 0;
    for(CheckersPiece& piece : bluePiecesOnBoard) {
        if(piece.isKing) {
            counter++;
        }
    }
    return counter;
}

int ImageState::countRedKingsOffBoard() {
    int counter = 0;
    for(CheckersPiece& piece : redPiecesOffBoard) {
        if(piece.isKing) {
            counter++;
        }
    }
    return counter;
}

int ImageState::countBlueKingsOffBoard() {
    int counter = 0;
    for(CheckersPiece& piece : bluePiecesOffBoard) {
        if(piece.isKing) {
            counter++;
        }
    }
    return counter;
}

bool ImageState::generateBoardstate(cv::Mat& img) {
    // Get points
    std::vector<std::vector<Point>> points;
    getPointsInImage(img, points);
    std::vector<Point> bluePoints = points.back();
    points.pop_back();
    std::vector<Point> redPoints = points.back();
    points.pop_back();
    // Clusterize
    std::vector<Cluster> blueClusters;
    clusterize(bluePoints, true, blueClusters);
    std::vector<Cluster> redClusters;
    clusterize(redPoints, false, redClusters);
    bool valid = generateBoardState(redClusters, blueClusters);
    isValidState = valid;
    // check if the move was legal here
    bool wasLegalMove = true;
    if(valid && wasLegalMove) {
        lastValidBoardState = boardState;
        return true;
    }
    else {
        return false;
    }
}

bool ImageState::generateBoardState(std::vector<Cluster>& redClusters, std::vector<Cluster>& blueClusters) {
    bool success = true;
    // Red Pieces
    for(Cluster& c : redClusters) {
        CheckersPiece cp;
        cp.isBlue = false;
        cp.isKing = c.isKing;
        cp.x = c.x;
        cp.y = c.y;
        cp.onBoard = edgeX[0] < cp.x && cp.x < edgeX[1] && edgeY[0] < cp.y && cp.y < edgeY[1];
        if(cp.onBoard)
            redPiecesOnBoard.push_back(cp);
        else
            redPiecesOffBoard.push_back(cp);
    }
    // Blue Pieces
    for(Cluster& c : blueClusters) {
        CheckersPiece cp;
        cp.isBlue = true;
        cp.isKing = c.isKing;
        cp.x = c.x;
        cp.y = c.y;
        cp.onBoard = edgeX[0] < cp.x && cp.x < edgeX[1] && edgeY[0] < cp.y && cp.y < edgeY[1];
        if(cp.onBoard)
            bluePiecesOnBoard.push_back(cp);
        else
            bluePiecesOffBoard.push_back(cp);
    }
    // Check positions on board
    char pos[8][8];
    for(int i = 0; i < 8; i++) {
        for(int j = 0; j < 8; j++) {
            pos[i][j] = '.';
        }
    }
    for(CheckersPiece& p : redPiecesOnBoard) {
        std::cout << p.x << ", " << p.y << "\n";
        cv::Point2i coord = getBoardPos(p);
        if(coord.x == -1) {
            // Invaloid coordinate
            success = false;
        }
        else if(pos[coord.x][coord.y] == '.') {
            if(p.isKing) {
                pos[coord.x][coord.y] = 'R';
            } 
            else {
                pos[coord.x][coord.y] = 'r';
            }
        } 
        else {
            // Two pieces on same spot
            success = false;
        }
    }
    for(CheckersPiece& p : bluePiecesOnBoard) {
        cv::Point2i coord = getBoardPos(p);
        if(coord.x == -1) {
            // Invalid coordinate
            success = false;
        }
        if(pos[coord.x][coord.y] == '.') {
            if(p.isKing) {
                pos[coord.x][coord.y] = 'B';
            } 
            else {
                pos[coord.x][coord.y] = 'b';
            }
        } 
        else {
            // Two pieces on same spot
            success = false;
        }
    }
    // update boardsate string
    for(int i = 0; i < 8; i++) {
        for(int j = 0; j < 8; j++) {
            boardState = boardState + pos[i][j];
        }
        boardState = boardState + "\n";
    }
    return success;
}

bool ImageState::alignCamera(cv::Mat& img) {
    int cornerWidth = 7;
    int cornerHeight = 7;
    cv::Size boardSize(cornerWidth, cornerHeight);
    std::vector<cv::Point2f> corners;
    bool found = cv::findChessboardCorners(img, boardSize, corners, cv::CALIB_CB_ADAPTIVE_THRESH + cv::CALIB_CB_FAST_CHECK);
    if(!found) {
        // Could not find checkers board
        std::cout << "Did not find grid\n";
        return false;
    }
    /*
    // Undistort
    std::vector<cv::Point3f> objPoints;
    std::vector<std::vector<cv::Point3f>> objPointsOuter;
    std::vector<std::vector<cv::Point2f>> cornersOuter;
    for(int i = 0; i < 7; i++) {
        for(int j = 0; j < 7; j++) {
            objPoints.push_back(cv::Point3f((float)j*10, (float)i*10, 0.0f));
        }
    }
    objPointsOuter.push_back(objPoints);
    cornersOuter.push_back(corners);
    cv::Mat k, d;
    std::vector<cv::Mat> rvecs, tvecs;
    //cv::calibrateCamera(objPointsOuter, cornersOuter, img.size(), k, d, rvecs, tvecs);
    cv::Mat newCameraMat = cv::getOptimalNewCameraMatrix(k, d, img.size(), 1, img.size());
    cv::Mat newImage;
    cv::undistort(img, newImage, k, d);
    */
    // Find grid size
    // Calculate average board width and height:
    float widthSum = 0;
    float heightSum = 0;
    float topSum = 0;
    float botSum = 0;
    float leftSum = 0;
    float rightSum = 0;
    for(int i = 0; i < boardSize.area(); i++) {
        // Check if corner is not on the right end
        if(i % 7 != 6) {
            widthSum += corners[i+1].x - corners[i].x;
        }
        else {
            rightSum += corners[i].x;
        }
        // Check if corner is on the left
        if(i % cornerWidth == 0) {
            leftSum += corners[i].x;
        }
        // Check if corner is not on the bottom
        if(i < boardSize.area() - cornerWidth) {
            heightSum += corners[i+cornerWidth].y - corners[i].y;     
        }
        else {
            botSum += corners[i].y;
        }
        // Check if corner is on the top
        if(i < cornerWidth) {
            topSum += corners[i].y;
        }
    }
    // The number of heights and widths is 1 less than the number of corners found per edge
    float avgWidth = widthSum / 42;
    float avgHeight = heightSum / 42;
    avgSquareHeight = (int)avgWidth;
    avgSquareWidth = (int)avgHeight;
    // Extrapolate out to board edges, x and y from coordinates to board are switched
    edgeY[0] = (int)(leftSum / cornerHeight - avgWidth);
    edgeY[1] = (int)(rightSum / cornerHeight + avgWidth);
    edgeX[0] = (int)(topSum / cornerWidth - avgHeight);
    edgeX[1] = (int)(botSum / cornerWidth + avgHeight);
    return true;
}

cv::Point2i ImageState::getBoardPos(CheckersPiece& p) {
    // Keep moving right or down until you get past the point
    cv::Point2i pos(0, 0);
    while(p.x > edgeX[0] + (pos.x+1)*avgSquareWidth) {
        pos.x++;
    }
    while(p.y > edgeY[0] + (pos.y+1)*avgSquareHeight) {
        pos.y++;
    }
    // Check distance, if too far, return -1
    // The calculated value is always larger than p.x or p.y
    // Must be in the middle 3/5 of the board
    int minWidth = avgSquareWidth / 5;
    int xDist = edgeX[0] + (pos.x+1)*avgSquareWidth - p.x;
    int yDist = edgeY[0] + (pos.y+1)*avgSquareHeight - p.y;
    if(xDist < minWidth || xDist > minWidth * 4) {
        // Piece is close to a border
        pos.x = -1;
        pos.y = -1;
    }
    return pos;
}

//////////////////////////////// Global Method Definitions //////////////////////////////////
void getPointsInImage(cv::Mat& img, std::vector<std::vector<Point>>& pointsList) {
    // Split RGB channels
    cv::Mat bgr[3];
    cv::split(img, bgr);
    cv::Mat blueChannel, greenChannel, redChannel;
    bgr[0].convertTo(blueChannel, CV_32SC1);
    bgr[1].convertTo(greenChannel, CV_32SC1);
    bgr[2].convertTo(redChannel, CV_32SC1);
    // Filter
    cv::Mat blueFilter = blueChannel - redChannel - greenChannel;
    cv::Mat redFilter = redChannel - blueChannel - greenChannel;
    cv::Mat yellowFilter = 0.5 * (redChannel + greenChannel) - 2*blueChannel;
    // Find points
    int kernelSize = KERNEL_SIZE;
    int kernelArea = kernelSize * kernelSize;
    double filterCutoff = 160 * kernelArea;
    std::vector<Point> bluePoints;
    std::vector<Point> redPoints;
    for(int i = 0; i < 1088; i += kernelSize) {
        for(int j = 0; j < 1920; j += kernelSize) {
            cv::Mat blueSquare = blueFilter(cv::Rect(j, i, kernelSize, kernelSize));
            cv::Mat redSquare = redFilter(cv::Rect(j, i, kernelSize, kernelSize));
            cv::Mat yellowSquare = yellowFilter(cv::Rect(j, i, kernelSize, kernelSize));
            // Check for blue
            if(cv::sum(blueSquare)[0] > filterCutoff) {
                Point p;
                p.x = i + kernelSize/2;
                p.y = j + kernelSize/2;
                p.type = BLUE;
                bluePoints.push_back(p);
            }
            // Check for red
            else if(cv::sum(redSquare)[0] > filterCutoff) {
                Point p;
                p.x = i + kernelSize/2;
                p.y = j + kernelSize/2;
                p.type = RED;
                redPoints.push_back(p);
            }
            // Check for yellow
            else if(cv::sum(yellowSquare)[0] > filterCutoff) {
                Point p;
                p.x = i + kernelSize/2;
                p.y = j + kernelSize/2;
                p.type = YELLOW;
                bluePoints.push_back(p);
                redPoints.push_back(p);
            }
        }
    }
    // Add to vector and return
    pointsList.push_back(redPoints);
    pointsList.push_back(bluePoints);
    return;
}

void clusterize(std::vector<Point>& pList, bool isBlue, std::vector<Cluster>& finalClusters) {
    std::vector<Cluster> clusters;
    for(Point& p : pList) {
        // Go through all clusters and stop after finding one
        bool found = false;
        for(Cluster& c : clusters) {
            if(c.checkRange(p)) {
                c.addPoint(p);
                found = true;
                break;
            }
        }
        if(!found) {
            Cluster newCluster;
            newCluster.addPoint(p);
            clusters.push_back(newCluster);
        }
    }
    // Finish cluster calculation
    for(Cluster& c : clusters) {
        c.finalize();
        if(c.isValid && c.isBlue == isBlue) {
            finalClusters.push_back(c);
        }
    }
    return;
}
